sound/soc/sof/pm.c

```diff
@@ -236,6 +236,11 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 			 ret);
 	}
 
+	/* release trace now */
+	ret = snd_sof_dma_trace_release(sdev);
+	if (ret < 0)
+		dev_warn(sdev->dev, "snd_sof_dma_trace_release failed with error: %d\n", ret);
+
 suspend:
```

Comment:
I'm not super excited about the spreading around of dtrace related calls, this is certainly going to make the SOF client support a no go..

Quoting the commit message for comment:
```
The recommended programming sequence for STOP trigger
for normal PCM streams is as follows:
1. Send the STOP_TRIGGER IPC to the FW
2. Stop the host DMA
3. Send the PCM_FREE IPC to do the DMA reset
4. Finally, perform stream reset and coupling of host/link DMAs.

But due to the lack of STOP and PCM_FREE IPC for trace, this sequence
is modified as follows:

1. Stop the host DMA for trace
2. Send the CTX_SAVE IPC to the FW for DMA reset
3. Perform stream reset and host/link DMA coupling in
snd_sof_dma_trace_release().
```
The **current** sequence is:
```
In `snd_sof_release_trace()`
1. Send the STOP_TRIGGER IPC to the FW (`snd_sof_dma_trace_trigger()`)
2. Stop the host DMA (`snd_sof_dma_trace_release()`)

in `sof_suspend()`
3. Send the CTX_SAVE IPC to the FW for DMA reset
```

The correct place for the 'Send the PCM_FREE IPC to do the DMA reset' and 'Finally, perform stream reset and coupling of host/link DMAs.' is within the `snd_sof_dma_trace_release()` so the sequence would be:
```
In `snd_sof_release_trace()`
1. Send the STOP_TRIGGER IPC to the FW (`snd_sof_dma_trace_trigger()`)
2. Stop the host DMA (`snd_sof_dma_trace_release()`)
3. Send the PCM_FREE IPC to do the DMA reset (`snd_sof_dma_trace_release()`)
4. Finally, perform stream reset and coupling of host/link DMAs. (`snd_sof_dma_trace_release()`)

in `sof_suspend()`
5. Send the CTX_SAVE IPC
```
Without moving dtrace related calls all around the place.

Reply:
Other thing I don't really see is how this will conflict with #3099?
It is doing almost similar thing, but in a different way.
Do you for example need to do all 4 steps for s0ix?

Reply:
@ujfalusi I guess the problem is the lack of the IPCs for dma-trace. But this will indeed complicate auxbus plus how we keep the DSP neutral. OTOH, the SOF interface should be a superset that has enough flexibility to support all vendors, so I think it's ok to extend the interface, but this needs to be generic enough.

Reply:
@kv2019i, @ranj063, correct me if I'm wrong, but the dma-trace on the firmware side is stopped/paused/off in response to the `SOF_IPC_PM_GATE` message if the BIT(4) is set (SOF_PM_NO_TRACE / HDA_PM_NO_DMA_TRACE).
The sequence proposed by this PR does not change the order regarding to dtrace DMA, it just moves the CTX_SAVE a bit earlier.

In light of this, the **current** sequence is:
```
In `snd_sof_release_trace()`
1. Send the STOP_TRIGGER IPC to the FW (`snd_sof_dma_trace_trigger()`)
2. Stop the host DMA (`snd_sof_dma_trace_release()`)
3. Send the CTX_SAVE IPC to the FW (unrelated to DMA)
4. Send SOF_IPC_PM_GATE to stop the DMA trace
```
and the presented sequence is:
```
1. Send the STOP_TRIGGER IPC to the FW (`snd_sof_dma_trace_trigger()`)
2. Send the CTX_SAVE IPC to the FW (unrelated to DMA)
3. Stop the host DMA (`snd_sof_dma_trace_release()`)
4. Send SOF_IPC_PM_GATE to stop the DMA trace
```
The issue boils down to a simple fact that SOF_IPC_PM_GATE is tasked to stop the DMA trace along with other unrelated things.
But If we first send SOF_IPC_PM_GATE with only BIT(4) set in flags (and if the pm_runtime_enable() is not refcounted on the firmware side) then we would only disable the trace and that could be embedded inside of the dtrace code itself to contain the code within it's scope.